File: 2019/Thang02_2018/17_02_2019/czero.cpp
#include <iostream>
#include <stdio.h>

using namespace std;
////////

const
    int maxk=32;

long long f[40][40];
long long n, k, res;
int n2;
////////

void create()
{
    f[0][0]=1;
    for (int i=1; i<=maxk; i++)
    {
        f[i][0]=1; f[i][i]=1;
        for (int j=1; j<i; j++)
            f[i][j]=f[i-1][j]+f[i-1][j-1];
    }
}
////////

void enter()
{
    //scanf("%lld%lld", &n, &k);
    long long x=n; n2=0;
    while (x>0)
    {
        n2++;
        x/=2;
    }
}
////////

int getb(int i, long long x)
{
    return (x>>(i-1)&1);
}

void process()
{
    res=0;
    if (k>32) return;
    int cou=0;
    for (int i=n2-1; i>=1; i--)
    {
        res+=f[i-1][k];
        if (getb(i, n)==0) cou++; else
        if (k>cou) res+=f[i-1][k-cou-1];
    }
    if (cou==k) res++;
}

int main()
{
    freopen("czero.inp", "r", stdin);
    freopen("czero.out", "w", stdout);
    create();
    while (scanf("%lld%lld", &n, &k)>0)
    {
        enter();
        process();
        cout << res << '\n';
    }
}

File: 2018/Thang10-2018/04-10-2018/czero.pas
Const   ginp='czero.inp';
        gout='czero.out';
        maxn=32;

Var     k,res,sl:longint;
        n:int64;
        f:array[0..maxn,0..maxn] of longint;

Procedure Create;
Var     i,j:longint;
Begin
        f[0,0]:=1;
        for i:=1 to maxn do
        begin
                f[i,0]:=1; f[i,i]:=1;
                for j:=1 to i-1 do
                        f[i,j]:=f[i-1,j]+f[i-1,j-1];
        end;
End;

Function Getb(x:int64;i:longint):longint;
Begin
        getb:=x shr (i-1) and 1;
End;

Procedure Process;
Var     i,cou,j:longint;
        x:int64;
Begin
        while not seekeof do
        begin
                readln(n,k);
                if k>31 then
                begin
                        writeln(0);
                        continue;
                end;
                x:=n; sl:=0;
                while x>0 do
                begin
                        inc(sl);
                        x:=x shr 1;
                end;
                res:=0; cou:=0;
                for i:=sl-1 downto 1 do
                begin
                        res:=res+f[i-1,k];
                        if getb(n,i)=0 then inc(cou) else
                        if k>cou then res:=res+f[i-1,k-cou-1];
                end;
                if cou=k then inc(res);
                writeln(res);
        end;
End;

Procedure SinhTest;
Var     i,cou,x:longint;
Begin
        n:=2724; k:=4;
        Assign(output,ginp); Rewrite(output);
        writeln(n,' ',k);
        res:=0;
        for i:=1 to n do
        Begin
                cou:=0; x:=i;
                while x>0 do
                begin
                        if x and 1=0 then inc(cou);
                        x:=x shr 1;
                end;
                if cou=k then inc(res);
        End;
        write(res);
        Close(output);
End;

Begin
        Assign(input,ginp); Assign(output,gout);
        Reset(input); Rewrite(output);
        Create;
        Process;
        Close(input); Close(output);
        //SinhTest;
End.
